packages/core/src/storages/request_queue_v2.ts

```diff
@@ -0,0 +1,988 @@
+import { REQUEST_QUEUE_HEAD_MAX_LIMIT } from '@apify/consts';
+import { ListDictionary, LruCache } from '@apify/datastructures';
+import { cryptoRandomObjectId } from '@apify/utilities';
+import type {
+    BatchAddRequestsResult,
+    Dictionary,
+    ProcessedRequest,
+    QueueOperationInfo,
+    RequestQueueClient,
+    RequestQueueInfo,
+    StorageClient,
+} from '@crawlee/types';
+import type { DownloadListOfUrlsOptions } from '@crawlee/utils';
+import { chunk, downloadListOfUrls, sleep } from '@crawlee/utils';
+import ow from 'ow';
+
+import type { AddRequestsBatchedOptions, AddRequestsBatchedResult } from './request_queue';
+import type { StorageManagerOptions } from './storage_manager';
+import { StorageManager } from './storage_manager';
+import {
+    API_PROCESSED_REQUESTS_DELAY_MILLIS,
+    MAX_QUERIES_FOR_CONSISTENCY,
+    QUERY_HEAD_BUFFER,
+    QUERY_HEAD_MIN_LENGTH,
+    STORAGE_CONSISTENCY_DELAY_MILLIS,
+    getRequestId,
+    purgeDefaultStorages,
+} from './utils';
+import { Configuration } from '../configuration';
+import { log } from '../log';
+import type { ProxyConfiguration } from '../proxy_configuration';
+import { Request } from '../request';
+import type { InternalSource, RequestOptions, Source } from '../request';
+
+// Double the limit of RequestQueue v1 (1_000_000) as we also store keyed by request.id, not just from uniqueKey
+const MAX_CACHED_REQUESTS = 2_000_000;
+
+/**
+ * When prolonging a lock, we do it for a minute from Date.now()
+ */
+const PROLONG_LOCK_BY_SECS = 60;
+
+/**
+ * This number must be large enough so that processing of all these requests cannot be done in
+ * a time lower than expected maximum latency of DynamoDB, but low enough not to waste too much memory.
+ * @internal
+ */
+const RECENTLY_HANDLED_CACHE_SIZE = 1000;
+
+interface RequestLruItem {
+    uniqueKey: string;
+    isHandled: boolean;
+    id: string;
+    hydrated: Request | null;
+    lockExpiresAt: number | null;
+}
+
+export interface RequestQueueV2OperationOptions {
+    /**
+     * If set to `true`:
+     *   - while adding the request to the queue: the request will be added to the foremost position in the queue.
+     *   - while reclaiming the request: the request will be placed to the beginning of the queue, so that it's returned
+     *   in the next call to {@apilink RequestQueueV2.fetchNextRequest}.
+     * By default, it's put to the end of the queue.
+     * @default false
+     */
+    forefront?: boolean;
+}
+
+/**
+ * @internal
+ */
+export interface RequestQueueV2OperationInfo extends QueueOperationInfo {
+    uniqueKey: string;
+}
+
+class RequestQueue {
+    log = log.child({ prefix: 'RequestQueueV2' });
```

Comment:
I would personally rename all the occurrences from `RequestQueueV2` to just `RequestQueue2`.

Reply:
I can do it if you want, but to me RequestQueue2 sounds more like a second instance than a new class